hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -452,6 +450,33 @@ public void initialize(URI name, Configuration originalConf)
 
   }
 
+  /**
+   * Test bucket existence in S3.
+   * When value of {@link Constants#S3A_BUCKET_PROBE is set to 0 by client,
+   * bucket existence check is not done to improve performance of
+   * S3AFileSystem initialisation. When set to 1 or 2, bucket existence check
+   * will be performed which is potentially slow.
+   * @throws IOException
+   */
+  private void doBucketProbing() throws IOException {
```

Comment:
add a RetryPolicy by looking @ the methods it calls and see what they do

We mustn't have a retry() calling operations which retry themselves anyway, as it explodes the #of retries which take place. It's ok to use once() round either of them, as the exception translating stuff is a no-op on already translated exceptions

Reply:
verifyBucketExists() and verifyBucketExistsV2() are the methods which are getting called from the doBucketProbing() method and I see they are already using invoked which has a retry policy set to TRY_ONCE_THEN_FAIL. Do we need to put explicit retry in this method? 
Also both these methods are annotated with RetryTranslated and the documentation of Retried say that if RetryTranslated is used, the called shouldn't perform another layer of retries.

Reply:
you just need to add the @RetryPolicy on the method based on the inner ones. it's not about actually doing the retries, just declare what it is for people looking at it.

The goal is that if we can keep those attributes accurate you just need to look at the method and determine the complete retrial policy that it has -all the way down.

This means we need to add the attribute to all methods, and keep an eye on them to make sure that they don't go invalid/out of date after changes underneath.

It's a shame we can't automate this -but the need to have them does force us to audit the code. Like you say: we mustn't retry around a retry -but we must have a retry somewhere above every non-retried operation.

Reply:
After reading some documentation, I understood what you mean to say here. So I have added @Retries.RetryTranslated annotation there. That means since the underling methods are retried and translated, the called shouldn't retry or translate again. Please correct me if I am wrong.

Reply:
yes. This is all just information for people looking at the code